File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        m ={}
        def dfs(node):
            if node in m:
                return m[node]
            clon = Node(node.val)
            m[node]=clon
            for x in node.neighbors:
                t = dfs(x)
                clon.neighbors.append(t)
            return clon
        return dfs(node) if node else None

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if(node == NULL) return NULL;
        map<Node *, Node *> m;
        function<Node*(Node*)> dfs =  [&](Node* p){
            if(m.find(p)!=m.end())
                return m[p];
            Node *t = new Node(p->val);
            m[p]=t;
            for(auto x:p->neighbors){
                Node* pp = dfs(x);
                t->neighbors.push_back(pp);
            }
            return t;
        };
        return dfs(node);
    }
};